### ISSUE TITLE
[cloud] C4: encrypted credentials via Fernet + user_credentials Postgres table

### ISSUE THREAD

**Issue by yadava5** [yadava5/jobtracker#21]
## Problem

Cloud deployments cannot use `keyring` (macOS Keychain is local-only). Gmail OAuth tokens and iCloud app-specific passwords currently live in the Keychain via `backend/jobtracker/credentials.py`. Without an alternative, the cloud backend cannot persist per-user secrets at all.

## Why it matters

Blocks C5 (Gmail web OAuth persistence), C7 (cron needs credentials per user to sync), and the iCloud flow on web.

## Real cause / current gap

- `backend/jobtracker/credentials.py` imports `keyring` at module top — the module is part of the desktop import graph but never loaded in cloud.
- No `user_credentials` table exists.
- `SECRET_ENCRYPTION_KEY` config field shipped in C1 but unused.

## Acceptance criteria

- [ ] Alembic migration adds `user_credentials` table: `(user_id UUID, kind TEXT, ciphertext BYTEA, nonce BYTEA, key_id TEXT, created_at, updated_at, PK(user_id, kind))`. `kind ∈ {gmail_oauth, icloud_mail}`.
- [ ] Split `backend/jobtracker/credentials.py` into a package:
  - `backend/jobtracker/credentials/__init__.py` — dispatcher selecting `desktop` or `cloud` based on `settings.deployment`.
  - `backend/jobtracker/credentials/desktop.py` — keyring-backed (existing behavior).
  - `backend/jobtracker/credentials/cloud.py` — Fernet-encrypted rows in `user_credentials`.
- [ ] Public API preserved: `get_gmail_credentials(user_id)`, `save_gmail_credentials(user_id, ...)`, `get_icloud_credentials(user_id)`, `save_icloud_credentials(user_id, ...)`, `has_*`, `delete_*`.
- [ ] Key rotation: `key_id` column references a `settings.secret_encryption_keys` dict keyed by id; default active key is id `"v1"`.
- [ ] Tests: `backend/tests/test_credentials_cloud.py` — round-trip encrypt→decrypt, key-rotation decryption of old ciphertext, missing-key error path.
- [ ] Existing desktop credential tests unchanged.
- [ ] Docs: `docs/WEB_MIGRATION.md` new section "Secret storage"; `docs/WEB_ARCHITECTURE.md` updated.

## Validation plan

```
.venv311/bin/pytest tests -q
python -c "from jobtracker.credentials import save_gmail_credentials, get_gmail_credentials; ..."  # cloud round-trip
# Verify DB row
psql $DATABASE_URL -c "SELECT user_id, kind, length(ciphertext) FROM user_credentials;"
```

## Dependencies

Blocked by: #16 (C2), C3 (user_id). Blocks: C5, C7.

## Out of scope

- Client secret rotation tooling beyond a key_id column.
- Per-user encryption key (current design: single `SECRET_ENCRYPTION_KEY` for all users; defense-in-depth via DB + RLS).

**Comment by yadava5** [comment]
## Implementation Spec (iteration-4 research, 2026-04-18)

# C4 — Encrypted credentials

## 1. File list (absolute paths)

**Delete / replace** (convert module to package):
- `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/credentials.py` — remove, rewrite as package.

**New files**:
- `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/credentials/__init__.py` — dispatcher.
- `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/credentials/types.py` — `GmailCredentials`, `ICloudCredentials` dataclasses (extracted from current `credentials.py:36-94`).
- `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/credentials/desktop.py` — keyring-backed implementation (lift-and-shift from `credentials.py:115-340`).
- `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/credentials/cloud.py` — Fernet + Postgres implementation.
- `/Users/ayush/Documents/Projects/jobtracker/backend/alembic/versions/<hash>_add_user_credentials.py` — DDL migration.
- `/Users/ayush/Documents/Projects/jobtracker/backend/alembic/versions/<hash>_user_credentials_rls.py` — RLS migration (Postgres-only).
- `/Users/ayush/Documents/Projects/jobtracker/backend/tests/test_credentials_cloud.py`.
- `/Users/ayush/Documents/Projects/jobtracker/backend/tests/test_credentials_dispatcher.py`.

**Modified**:
- `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/database/models.py` — add `UserCredential` SQLModel (cloud only; `table=True` when `settings.deployment == "cloud"` or always-declared with composite PK).
- `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/config.py` — extend `secret_encryption_key` into `secret_encryption_keys: dict[str, str]` (backward-compat: single-key maps to `{"v1": <key>}`).
- `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/api/auth.py` — every call-site currently at lines 111-112, 157, 182, 212, 241, 287, 300, 309, 326, 355 takes `user_id: UUID = Depends(current_user)` and threads it into the credential calls.
- `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/api/sync.py:97-98` — pass `user_id`.
- `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/email_clients/gmail.py:37-43, 118, 123, 206, 222, 637` and `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/email_clients/icloud.py:35, 119, 123, 140` — accept `user_id` on construction and thread to all credential calls.
- `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/services/sync.py:25, 138-139` — same.
- `/Users/ayush/Documents/Projects/jobtracker/backend/jobtracker/main.py:205-210` — desktop health probe: pass sentinel `LOCAL_USER_ID = UUID(int=0)`.

## 2. Public-API contract (stable, both backends)

Every function MUST accept `user_id: UUID` as the first positional arg. Return types unchanged from today.

```
save_gmail_credentials(user_id: UUID, credentials: GmailCredentials) -> bool
get_gmail_credentials(user_id: UUID) -> Optional[GmailCredentials]
delete_gmail_credentials(user_id: UUID) -> bool
update_gmail_access_token(user_id: UUID, access_token: str, token_expiry: datetime) -> bool
has_gmail_credentials(user_id: UUID) -> bool
save_icloud_credentials(user_id: UUID, credentials: ICloudCredentials) -> bool
get_icloud_credentials(user_id: UUID) -> Optional[ICloudCredentials]
delete_icloud_credentials(user_id: UUID) -> bool
has_icloud_credentials(user_id: UUID) -> bool
save_gmail_client_secret(user_id: UUID, client_secret_json: dict) -> bool   # desktop only; cloud raises NotImplementedError
get_gmail_client_secret(user_id: UUID) -> Optional[dict]                    # cloud returns env-sourced dict
clear_all_credentials(user_id: UUID) -> bool
```

Data classes (`GmailCredentials`, `ICloudCredentials`) are preserved byte-for-byte from `credentials.py:36-94`.

## 3. Dispatcher strategy

`credentials/__init__.py` does **lazy-imported module re-export**, not function wrappers, to avoid double-decorating log lines and to let tests monkeypatch the backend directly:

```python
from jobtracker.config import settings
if settings.deployment == "cloud":
    from jobtracker.credentials.cloud import *  # noqa
else:
    from jobtracker.credentials.desktop import *  # noqa
from jobtracker.credentials.types import GmailCredentials, ICloudCredentials
```

`desktop.py` keeps `import keyring` at top. `cloud.py` keeps `from cryptography.fernet import MultiFernet, Fernet` at top. Neither backend is imported on the other deployment, so `keyring` never loads on Vercel.

## 4. Alembic migration — `user_credentials`

```sql
CREATE TABLE user_credentials (
  user_id     UUID        NOT NULL,
  kind        TEXT        NOT NULL CHECK (kind IN ('gmail_oauth','icloud_mail')),
  ciphertext  BYTEA       NOT NULL,
  nonce       BYTEA       NOT NULL,       -- reserved for future AEAD; Fernet embeds IV, store b''
  key_id      TEXT        NOT NULL DEFAULT 'v1',
  created_at  TIMESTAMPTZ NOT NULL DEFAULT now(),
  updated_at  TIMESTAMPTZ NOT NULL DEFAULT now(),
  PRIMARY KEY (user_id, kind),
  CONSTRAINT fk_user_credentials_user FOREIGN KEY (user_id)
    REFERENCES auth.users(id) ON DELETE CASCADE
);
CREATE INDEX idx_user_credentials_kind ON user_credentials(kind);
```

`op.execute` everything that references `auth.users` is wrapped with a dialect guard (`if op.get_bind().dialect.name == "postgresql"`) so SQLite desktop runs skip the FK. Cloud-only execution is fine because C4's cloud backend never writes on SQLite.

**RLS migration** (second file):
```sql
ALTER TABLE user_credentials ENABLE ROW LEVEL SECURITY;
CREATE POLICY user_credentials_owner ON user_credentials
  USING (user_id = auth.uid()) WITH CHECK (user_id = auth.uid());
```
Both migrations gate with `if bind.dialect.name == "postgresql"`.

## 5. Cloud implementation notes (`cloud.py`)

- Use `MultiFernet` seeded from `settings.secret_encryption_keys` ordered with active key (id `v1` by default) first. Encrypt uses the first (active) key only; decrypt iterates all. After successful decrypt with a non-active key, re-encrypt and `UPDATE` the row (automatic key rotation).
- Sync DB calls. The existing `jobtracker.database.get_session` is async; call it from async functions. Public API becomes `async def` in cloud, but desktop stays sync. **Resolution**: make **all** public functions `async def` so API surface is uniform; desktop wraps synchronous keyring calls and awaits trivially. This is a breaking change at call sites — every `get_gmail_credentials()` becomes `await get_gmail_credentials(user_id)`. Acceptable because issue scope already requires touching all call-sites to add `user_id`.
- Missing-key path: `KeyError("no encryption key active")` bubbled as `RuntimeError` at boot; per-request decrypt failure returns `None` with a warning log.

## 6. Route specs

No new routes. C4 is infra. Existing `/auth/*` routes retain path/method/shape; only add `user_id` dependency injection.

## 7. Test plan — `backend/tests/test_credentials_cloud.py`

Key asserts (each a separate test):
- round-trip `save_gmail_credentials` → `get_gmail_credentials` equals original dataclass.
- `save_gmail_credentials` writes exactly 1 row per `(user_id, kind)`; second save updates, not duplicates.
- rotation: write with key `v1`, add `v2` active, read → decrypts via `v1`, row re-encrypted with `v2` on next cycle.
- missing-key error: clear `settings.secret_encryption_keys` → constructor raises `RuntimeError`.
- two users' rows are isolated: user A cannot read user B's row (verified via SQL directly; RLS is a Postgres-integration test left as xfail on SQLite).
- `delete_gmail_credentials` returns False when row absent (parity with desktop).
- `has_gmail_credentials` maps to `SELECT 1 FROM ... LIMIT 1`, no decryption.

## 8. Dependency on C3

C3 ships `jobtracker.auth.supabase_jwt.current_user` → `Depends(current_user) -> UUID`. C4 consumes it in `api/auth.py` refactor:
```python
@router.get("/status")
async def get_auth_status(user_id: UUID = Depends(current_user)) -> AuthStatusResponse:
    creds = await get_gmail_credentials(user_id)
    ...
```
For desktop, `require_user` returns `LOCAL_USER_ID` (a fixed sentinel UUID). Same decorator, different DI wiring, per C3's `require_user()` helper.

## 9. Risks and mitigations

1. **Key loss = total data loss for credentials.** Mitigation: `settings.secret_encryption_keys` is a dict; operators add `v2` before retiring `v1`. Add `scripts/rotate_fernet_keys.py` (out-of-scope for this issue but documented in `WEB_MIGRATION.md`).
2. **Async/sync surface shift breaks desktop.** Mitigation: land as two PRs — first introduce `async def` stubs that internally run sync keyring calls via `asyncio.to_thread`; only then split the module. Every desktop test gets an `await` but logic is unchanged.
3. **RLS skipped on SQLite hides bugs.** Mitigation: add an integration test marked `@pytest.mark.postgres` that runs against a Supabase shadow DB in CI when `SUPABASE_TEST_URL` is set.

## 10. Acceptance-criteria mapping

| AC bullet | Deliverable |
|---|---|
| Alembic adds `user_credentials` | `versions/<hash>_add_user_credentials.py` — §4 |
| Package split (desktop/cloud) | `credentials/{__init__,desktop,cloud,types}.py` — §1 |
| Public API preserved | Signature table — §2 |
| Key rotation via `key_id` | `MultiFernet` decrypt-and-re-encrypt — §5 |
| `test_credentials_cloud.py` coverage | Round-trip, rotation, missing-key — §7 |
| Existing desktop tests unchanged | `desktop.py` is lift-and-shift; only `user_id` param added — §3 |
| Docs: WEB_MIGRATION Secret storage | New `## Secret storage` section; 30-line key-rotation runbook |
| Docs: WEB_ARCHITECTURE updated | Add Mermaid diagram of encryption path cloud→Postgres |

## Doc updates

- `docs/WEB_MIGRATION.md`: new `## Secret storage` section (how to generate `JOBTRACKER_SECRET_ENCRYPTION_KEY`, how to rotate, what happens on key loss).
- `docs/WEB_ARCHITECTURE.md`: append "Credentials" subsection after Auth (references C3), include dispatcher decision tree and Fernet flow.
- `docs/API_SPEC.md`: no route changes; add note "cloud requires `Authorization: Bearer <jwt>` on all `/auth/*` routes (C3)".

**Comment by yadava5** [comment]
Closed by #31 — merged to integration/web-migration.
